File: packages/python/examples/existing_instance.py
"""
Examples of using HogTyped with an existing PostHog instance
"""

import posthog
from posthog_generated import hogtyped, PostHog

# ============================================
# OPTION 1: Pass instance to constructor
# ============================================

# Your existing PostHog client (already configured)
my_posthog = posthog.Client(
    api_key='phc_your_api_key',
    host='https://app.posthog.com',
    debug=True,
    on_error=lambda e: print(f"PostHog error: {e}")
)

# Create a new HogTyped wrapper using YOUR instance
my_analytics = PostHog(posthog_instance=my_posthog)

# Now use type-safe events with your existing PostHog setup
my_analytics.capture(
    distinct_id='user-123',
    event='simple_event',
    properties={
        'name': 'Product View',
        'count': 1
    }
)

# ============================================
# OPTION 2: Use set_instance() method
# ============================================

# Start with the default singleton
# Later, update it to use your configured instance
hogtyped.set_instance(my_posthog)

# Now it uses your instance
hogtyped.capture(
    distinct_id='user-123',
    event='complex_event',
    properties={
        'id': '123',
        'status': 'completed'
    }
)

# ============================================
# OPTION 3: Multiple PostHog projects
# ============================================

# You might have multiple PostHog projects
project_a = posthog.Client(
    api_key='phc_project_a_key',
    host='https://app.posthog.com'
)

project_b = posthog.Client(
    api_key='phc_project_b_key',
    host='https://eu.posthog.com'
)

# Create separate wrappers for each
analytics_a = PostHog(posthog_instance=project_a)
analytics_b = PostHog(posthog_instance=project_b)

# Track to different projects with type safety
analytics_a.capture(
    distinct_id='user-123',
    event='simple_event',
    properties={
        'name': 'Event for Project A',
        'count': 1
    }
)

analytics_b.capture(
    distinct_id='user-456',
    event='simple_event',
    properties={
        'name': 'Event for Project B',
        'count': 2
    }
)

# ============================================
# OPTION 4: Django integration
# ============================================

# In your Django settings
from django.conf import settings

# Your existing PostHog setup in Django
django_posthog = posthog.Client(
    api_key=settings.POSTHOG_API_KEY,
    host=settings.POSTHOG_HOST
)

# Create wrapper with Django's instance
django_analytics = PostHog(posthog_instance=django_posthog)

# In your views
def track_user_action(user, action):
    django_analytics.capture(
        distinct_id=str(user.id),
        event='user_action',
        properties={
            'action': action,
            'email': user.email
        }
    )

# ============================================
# OPTION 5: Flask integration
# ============================================

from flask import Flask, g

app = Flask(__name__)

# Initialize PostHog once for your Flask app
flask_posthog = posthog.Client(
    api_key='phc_flask_key',
    host='https://app.posthog.com'
)

# Create wrapper
flask_analytics = PostHog(posthog_instance=flask_posthog)

@app.before_request
def before_request():
    g.analytics = flask_analytics

@app.route('/api/action')
def track_action():
    g.analytics.capture(
        distinct_id='user-123',
        event='api_call',
        properties={'endpoint': '/api/action'}
    )
    return {'status': 'ok'}

# ============================================
# OPTION 6: Testing with mock instance
# ============================================

# For unit tests
class MockPostHog:
    def capture(self, **kwargs):
        print(f"Mock capture: {kwargs}")

    def identify(self, **kwargs):
        print(f"Mock identify: {kwargs}")

    def feature_enabled(self, key, distinct_id):
        return True

# Use mock in tests
test_analytics = PostHog(posthog_instance=MockPostHog())
test_analytics.capture(
    distinct_id='test-user',
    event='simple_event',
    properties={'name': 'Test', 'count': 1}
)

# ============================================
# OPTION 7: Async/Celery workers
# ============================================

from celery import Celery

# Configure PostHog for background workers
worker_posthog = posthog.Client(
    api_key='phc_worker_key',
    host='https://app.posthog.com',
    sync_mode=False  # Use async mode for better performance
)

worker_analytics = PostHog(posthog_instance=worker_posthog)

@app.task
def process_user_action(user_id, action):
    worker_analytics.capture(
        distinct_id=user_id,
        event='background_action',
        properties={'action': action}
    )

# ============================================
# OPTION 8: Access underlying instance
# ============================================

# Create wrapper
analytics = PostHog(posthog_instance=my_posthog)

# Use type-safe wrapper
analytics.capture(
    distinct_id='user-123',
    event='button_clicked',
    properties={'button_id': 'submit'}
)

# Access underlying PostHog for advanced features
analytics.posthog.flush()  # Force flush events
analytics.posthog.debug = True  # Enable debug mode
analytics.posthog.disabled = False  # Enable/disable tracking

# Use any other posthog-python feature directly
analytics.posthog.capture(
    distinct_id='user-123',
    event='$pageview',
    properties={'$current_url': 'https://example.com'}
)

# ============================================
# WHY THIS IS USEFUL
# ============================================

"""
1. **Existing Setup**: Don't need to change your PostHog initialization
2. **Multiple Projects**: Can use different wrappers for different projects
3. **Framework Integration**: Works with Django, Flask, FastAPI setups
4. **Testing**: Can inject mock instances for testing
5. **Background Workers**: Can use with Celery, RQ, etc.
6. **Custom Config**: Keep all your custom PostHog configuration
7. **Advanced Features**: Access underlying instance for features not in wrapper
"""

File: packages/js/examples/existing-instance.ts
/**
 * Examples of using HogTyped with an existing PostHog instance
 */

import posthog from 'posthog-js';
import hogtyped, { Analytics } from '../dist/posthog.generated';

// ============================================
// OPTION 1: Pass instance to constructor
// ============================================

// Your existing PostHog setup (already configured)
const myPostHog = posthog.init('phc_your_api_key', {
  api_host: 'https://app.posthog.com',
  persistence: 'localStorage',
  autocapture: true,
  capture_pageview: true,
  capture_pageleave: true,
  // ... your custom config
});

// Create a new HogTyped wrapper using YOUR instance
const myAnalytics = new Analytics(myPostHog);

// Now use type-safe events with your existing PostHog setup
myAnalytics.capture('simple_event', {
  name: 'Product View',
  count: 1
});

// ============================================
// OPTION 2: Use setInstance() method
// ============================================

// Start with the default singleton
// Later, update it to use your configured instance
hogtyped.setInstance(myPostHog);

// Now it uses your instance
hogtyped.capture('complex_event', {
  id: '123',
  status: 'completed'
});

// ============================================
// OPTION 3: Multiple PostHog projects
// ============================================

// You might have multiple PostHog projects
const projectA = posthog.init('phc_project_a_key', {
  api_host: 'https://app.posthog.com',
  persistence_name: 'project_a'
});

const projectB = posthog.init('phc_project_b_key', {
  api_host: 'https://eu.posthog.com',
  persistence_name: 'project_b'
});

// Create separate wrappers for each
const analyticsA = new Analytics(projectA);
const analyticsB = new Analytics(projectB);

// Track to different projects with type safety
analyticsA.capture('simple_event', {
  name: 'Event for Project A',
  count: 1
});

analyticsB.capture('simple_event', {
  name: 'Event for Project B',
  count: 2
});

// ============================================
// OPTION 4: Next.js / SSR with client instance
// ============================================

// In your app initialization (e.g., _app.tsx)
import { useEffect, useState } from 'react';

function MyApp() {
  const [analytics, setAnalytics] = useState<Analytics | null>(null);

  useEffect(() => {
    // Only initialize on client side
    if (typeof window !== 'undefined') {
      const ph = posthog.init('phc_your_key', {
        loaded: (posthog) => {
          // Create wrapper with the loaded instance
          const wrapper = new Analytics(posthog);
          setAnalytics(wrapper);
        }
      });
    }
  }, []);

  // Use analytics safely
  if (analytics) {
    analytics.capture('page_view', { path: window.location.pathname });
  }
}

// ============================================
// OPTION 5: Testing with mock instance
// ============================================

describe('Analytics Tests', () => {
  let mockPostHog: any;
  let testAnalytics: Analytics;

  beforeEach(() => {
    // Create a mock PostHog instance
    mockPostHog = {
      capture: jest.fn(),
      identify: jest.fn(),
      getFeatureFlag: jest.fn().mockReturnValue('variant-a')
    };

    // Use mock in wrapper
    testAnalytics = new Analytics(mockPostHog);
  });

  test('should track events with correct properties', () => {
    testAnalytics.capture('simple_event', {
      name: 'Test Event',
      count: 5
    });

    expect(mockPostHog.capture).toHaveBeenCalledWith(
      'simple_event',
      { name: 'Test Event', count: 5 }
    );
  });
});

// ============================================
// OPTION 6: Access underlying instance
// ============================================

// Access the underlying posthog-js for advanced features
const analytics = new Analytics(myPostHog);

// Use type-safe wrapper
analytics.capture('button_clicked', {
  button_id: 'submit'
});

// Access underlying instance for session replay
analytics.posthog.startSessionRecording();

// Check feature flags directly
const flags = analytics.posthog.getAllFlags();

// Use any other posthog-js feature
analytics.posthog.debug();

// ============================================
// WHY THIS IS USEFUL
// ============================================

/*
1. **Existing Setup**: Don't need to change your PostHog initialization
2. **Multiple Projects**: Can use different wrappers for different projects
3. **Custom Config**: Keep all your custom PostHog configuration
4. **Testing**: Can inject mock instances for testing
5. **SSR**: Can handle client-only initialization properly
6. **Advanced Features**: Access underlying instance for features not in wrapper
*/